cmd/backport-verifier/server.go

```diff
@@ -0,0 +1,188 @@
+			return
+		}
+	}
+
+	invalidCommits := map[string]string{}
+	validCommits := map[string]string{}
+	upstreamPullsByCommit := map[string]int{}
+	errorsByCommit := map[string]string{}
+	for _, commit := range commits {
+		parts := upstreamPullRe.FindStringSubmatch(commit.Commit.Message)
+		if len(parts) != 2 {
+			invalidCommits[commit.Commit.SHA] = "does not specify an upstream backport in the commit message"
+			continue
+		}
+		pr, err := strconv.Atoi(parts[1])
+		if err != nil {
+			// based on the regex this should not happen ...
+			logger.WithError(err).Warn("Failed to parse PR as integer")
+			errorsByCommit[commit.Commit.SHA] = fmt.Sprintf("failed to parse PR identifier: %s", err.Error())
+		} else {
+			upstreamPullsByCommit[commit.Commit.SHA] = pr
+		}
+	}
+
+	for commit, pull := range upstreamPullsByCommit {
+		prMention := fmt.Sprintf("the upstream PR [%s/%s#%d](https://github.com/%s/%s/pull/%d)", upstreamOrg, upstreamRepo, pull, upstreamOrg, upstreamRepo, pull)
+		pr, err := s.ghc.GetPullRequest(upstreamOrg, upstreamRepo, pull)
+		if err != nil {
+			if github.IsNotFound(err) {
+				invalidCommits[commit] = prMention + " does not exist"
+			} else {
+				logger.WithError(err).Warn("Failed to get upstream PR")
+				errorsByCommit[commit] = fmt.Sprintf("failed to fetch upstream PR: %s", err.Error())
+			}
+			continue
+		}
+		if !pr.Merged {
+			invalidCommits[commit] = prMention + " has not yet merged"
+		} else {
+			validCommits[commit] = prMention + " has merged"
+		}
+	}
+
+	desired := invalidBackportsLabel
+	verb := "could not"
+	if len(errorsByCommit) == 0 && len(invalidCommits) == 0 {
+		desired = validBackportsLabel
+		verb = "could"
+	}
+	ensureLabels(s.ghc, l, desired, org, repo, num)
+
+	message := fmt.Sprintf("@%s: the contents of this pull request %s be automatically validated.", user, verb)
+	for _, item := range []struct {
+		qualifier string
+		data      map[string]string
+	}{
+		{"are valid", validCommits},
+		{"are invalid", invalidCommits},
+		{"could not be processed", errorsByCommit},
+	} {
+		if len(item.data) > 0 {
+			var formatted []string
+			for commit, why := range item.data {
+				formatted = append(formatted, fmt.Sprintf(" - %s: %s", commit, why))
+			}
+			sort.Strings(formatted)
+			message = fmt.Sprintf("%s\n\nThe following commits %s:\n%s", message, item.qualifier, strings.Join(formatted, "\n"))
+		}
+	}
+	if commentErr := s.ghc.CreateComment(org, repo, num, message); commentErr != nil {
+		logger.WithError(commentErr).Warn("couldn't respond to user")
+	}
+}
+
+func ensureLabels(client githubClient, l *logrus.Entry, desired string, org, repo string, num int) {
+	var unwanted string
+	if desired == validBackportsLabel {
+		unwanted = invalidBackportsLabel
+	} else {
+		unwanted = validBackportsLabel
+	}
+	if err := client.AddLabel(org, repo, num, desired); err != nil {
+		l.WithError(err).Warn("could not add label", err)
+	}
+	if err := client.RemoveLabel(org, repo, num, unwanted); err != nil {
```

Comment:
Is it safe to assume that this already exists? Shouldn't we tolerate a 404 here?

Reply:
I thought the client did internally, let me check

Reply:
yeah it does